### PR TITLE
fix: ebs csi driver addon (jx#8575)

### PR DESCRIPTION
**Pull request by tgelpi**
#### Description 
Building/upgrading AWS to  v1.23 and greater versions now require a container storage interface (CSI) driver. 

[The Kubernetes in-tree to container storage interface (CSI) volume migration feature was enabled in K8S version 1.23.](https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html) This feature enables the replacement of existing Kubernetes in-tree storage plugins for Amazon EBS with a corresponding Amazon EBS CSI driver. 

An add-on is software that provides supporting operational capabilities to Kubernetes applications, but is not specific to the application.  [EKS Addons Guide ](https://docs.aws.amazon.com/eks/latest/userguide/eks-add-ons.html) 

This fix introduces Terraform resources to build and manage the Amazon EKS Add-on for the Amazon Elastic Block Store Container Storage Interface **(Amazon EBS CSI)** 

#### Which issue this PR fixes
jenkins-x/jx#8575

#### Release notes
The proposed configuration may require multiple `terraform apply` commands executed due to Terraform resource dependency limitations.   Currently, the first apply execution can result with the following error message:
```
Error: Kubernetes cluster unreachable: the server has asked for the client to provide credentials
```
Performing another apply command should resolve the issue.

### CODE REVIEW

**Comment by tgelpi** [comment]
I think I resolved the "_cluster unreachable_" issue. Removed a dependency that I added for the **helm_release.jx-git-operator** resource.  Will update the **enable_ebs** branch.

File: main.tf [hunk 1]
```diff
@@ -120,6 +122,7 @@ module "vault" {
   use_vault      = var.use_vault
   region         = var.region
   enable_acl     = var.enable_acl
+  depends_on     = [module.cluster.eks]
```
Comment (msvticket) [inline main.tf:125]: Why is this needed?

File: variables.tf [hunk 2]
```diff
@@ -379,7 +379,7 @@ variable "allowed_spot_instance_types" {
 variable "enable_worker_groups_launch_template" {
   description = "Flag to enable Worker Group Launch Templates"
   type        = bool
-  default     = false
+  default     = true
```
Comment (msvticket) [inline variables.tf:382]: Why is this default changed?

File: README.md [hunk 3]
```diff
@@ -545,6 +546,55 @@ module "eks-jx" {
...
+  count             = var.enable_ebscsi_driver ? 1 : 0
+  cluster_name      = var.cluster_name
+  addon_name        = var.ebs_addon.name
+  addon_version     = var.ebs_addon.version
+  resolve_conflicts = "OVERWRITE"
+  depends_on = [
+    null_resource.kubeconfig
+  ]
+}
+```
```
Comment (msvticket) [inline README.md:596]: I think this is only confusing to put in the documentation. A user that want to see the implementation can just look at the code.

File: modules/cluster/variables.tf [hunk 4]
```diff
@@ -447,3 +447,22 @@ variable "enable_acl" {
   description = "Flag to enable ACL instead of bucket ownership for S3 storage"
   type        = bool
 }
+
+variable "enable_ebscsi_driver" {
+  description = "Flag to enable or disable EBS CSI driver addon"
+  type        = bool
+  default     = true
```
Comment (msvticket) [inline modules/cluster/variables.tf:454]: I don't think this should be true by default since that would change the behaviour for an existing user. But you should add enable_ebscsi_driver=true to https://github.com/jx3-gitops-repositories/jx3-terraform-eks when this is merged. Probably the version needs to be a variable in jx3-terraform-eks.

File: modules/cluster/main.tf [hunk 5]
```diff
@@ -128,7 +128,8 @@ module "eks" {
   node_groups = !var.enable_worker_group ? local.node_groups_extended : {}
 
   workers_additional_policies = [
-    "arn:${data.aws_partition.current.partition}:iam::aws:policy/AmazonEC2ContainerRegistryPowerUser"
+    "arn:${data.aws_partition.current.partition}:iam::aws:policy/AmazonEC2ContainerRegistryPowerUser",
+    var.enable_ebscsi_driver ? "arn:${data.aws_partition.current.partition}:iam::aws:policy/service-role/AmazonEBSCSIDriverPolicy" : ""
```
Comment (msvticket) [inline modules/cluster/main.tf:132]: Don't add the permissions to the workers. Instead an IAM role should be created (see irsa.tf) and attached to the addon using service_account_role_arn.

**Review by msvticket — CHANGES_REQUESTED** [review]
Thank you for making a stab a this

File: main.tf (hunk 1 above)
Comment (tgelpi) [reply main.tf:125]: The EBS add-on would result in a DEGRADED status. I suspected it was due to vault resources creation contention. I struggled with trying to determine if this was the case but it did seem to occur only when using Vault secrets and not ASM. I could retest this to confirm if it is required if you think it would be helpful.

File: variables.tf (hunk 2 above)
Comment (tgelpi) [reply variables.tf:382]: The default configuration with this flag set to false would not work. For now, I'm not sure how to configure the add-on with this default setting.

File: README.md (hunk 3 above)
Comment (tgelpi) [reply README.md:596]: I agree. I can remove this.

File: modules/cluster/main.tf (hunk 5 above)
Comment (tgelpi) [reply modules/cluster/main.tf:132]: I can work on this.

File: modules/cluster/variables.tf (hunk 4 above)
Comment (tgelpi) [reply modules/cluster/variables.tf:454]: I can configure it that way. Should I also add _enable_worker_groups_launch_template=true_  [ https://github.com/jx3-gitops-repositories/jx3-terraform-eks](https://github.com/jx3-gitops-repositories/jx3-terraform-eks) when this is merged?

File: main.tf (hunk 1 above)
Comment (tgelpi) [reply main.tf]: Not sure what I was experiencing earlier but the dependency isn't necessary. Removed the statement and tested on new builds in both Vault and ASM with success.

**Comment by msvticket** [comment]
> The following variable must be set to true for initial creation of environment:
> 
> ```
> enable_worker_groups_launch_template = true
> ```
> 
> The launch template variable can be set back to false after initial creation of environment.

Why is that?

**Comment by tgelpi** [comment]
If the `enable_worker_groups_launch_template` variable is set to false the current version and proposed versions do not complete a new build . The `kubectl get nodes` command  returns empty. You can set the variable to true, complete the build, then set the variable to false but not sure how stable the new environment is.

File: modules/cluster/outputs.tf [hunk 6]
```diff
@@ -94,3 +94,9 @@ output "pipeline_viz_iam_role" {
   value       = module.iam_assumable_role_pipeline_visualizer.this_iam_role_name
   description = "The IAM Role that the pipeline visualizer pod will assume to authenticate"
 }
+
+output "ebscsi_addon_iam_role" {
```
Comment (msvticket) [inline modules/cluster/outputs.tf]: There seem to be a typo here so it doesn't match the output of the parent module
```suggestion
output "ebs_csi_addon_iam_role" {
```

**Review by msvticket — CHANGES_REQUESTED** [review]
I have now tried to create a cluster using the current state of https://github.com/jx3-gitops-repositories/jx3-terraform-eks

Indeed, the created the nodes didn't join the cluster. I didn't investigate that further but instead tried setting  `enable_worker_groups_launch_template = true` the nodes created with this mechanism did join the cluster.

But if you set `enable_worker_groups_launch_template = true` you should at the same time set `enable_worker_group = false`. Otherwise you will end up with useless nodes running. For now I still think these changes should be done in jx3-terraform-eks.

**Review by msvticket — APPROVED** [review]

File: modules/cluster/outputs.tf (hunk 6 above)
Comment (tgelpi) [reply modules/cluster/outputs.tf]: The new ebscsi role introduced is implemented with the TF module `iam-role-for-service-accounts-eks` and not `iam-assumable-role-with-oidc` like the others.  This TF module seems more straightforward to implement and don't know if there's limits rather thatn using the assumable-role version. If you have concerns about doing this please let me know. With this  different module I'm struggling with getting the correct syntax for the appropriate output variables.  For now I've removed the output statements under this branch (enable_ebs) and will create a different branch to test output var tests.  Not sure if this new ebs_csi role needs to have output variables defined. I can also test with `enable_worker_group = false` and attempt to documents these nuances
